spicy/toolchain/src/compiler/visitors/validator.cc

```diff
@@ -335,6 +335,38 @@ struct PreTransformVisitor : public hilti::visitor::PreOrder<void, PreTransformV
         }
     }
 
+    void operator()(const hilti::expression::Keyword& n, position_t p) {
+        switch ( n.kind() ) {
+            case hilti::expression::keyword::Kind::DollarDollar: {
+                // We want to match ASTs where a field hook on a `void` field
+                // has `$$` as any descendant (realistically: unlikely as a
+                // direct child). Below we instead check that a `$$` has a hook
+                // and a field as parent where the field has a `void` type
+                // which is not the same, but seems close enough for now.
+                //
+                // TODO(bbannier): Consider performing this test as part of the
+                // `Hook` validation. This would require a mechanism to filter
+                // all descendants of a node by type.
+
+                auto hook = p.findParent<spicy::Hook>();
+                auto field = p.findParent<type::unit::item::Field>();
```

Comment:
This looks like it'll work only for inline hooks, not ones added externally. Given the test below, it wouldn't catch this one:

```
hook void Test::X() { print $$; }
```

This is tricky to solve because the external hook won't be tied to the Unit anymore (but to the HILTI-side struct that's been created from the unit), so there's no field. What we could do is check if the hook's parameter list includes a parameter called `__dd`, as that's what ``$$`` will be mapped to. Not exactly nice, but another artifact of the current resolver process where we mix HILTI and Spicy AST processing into the just a single loop. 

We could also just punt and skip the check, unlikely that users will ever run into this.

Reply:
I added a patch to this PR which rejects `void` fields with names so that it would become impossible to declared external field hooks on `void` fields.

We still have an issue with `$$` being accepted in hooks which do not provide it, e.g., `%done` hooks. I took a stab at adding validation for that as well along the lines you suggested above, but it looks like I cannot see the parameters of such hooks, e.g., slightly modifying the lines changed here to also include

```.cpp
                const auto& parameters = hook->get().type().parameters();
                if ( parameters.end() ==
                     std::find_if(parameters.begin(), parameters.end(), [](auto&& p) { return p.id() == ID("__dd"); }) )
                    error("$$ is undefined", p);
```

Here I _always_ see an empty parameter list for any of the following hook impls,

```.ruby
module Test;

public type X = unit {
    a: uint8 { print $$; } # Okay.
    on %done { print $$; } # Should fail: `$$` not available.
};

on Test::X::%done { print $$; } # Should fail: `$$` not available.
```

We probably need a broader fix which properly resolves `$$` in hooks instead of magically pasting in code without checking whether the value is defined; this should work for different kind of hooks and e.g., also for hooks on `void` fields which have a different signature. If we did that we wouldn't even need the fix here, but instead could use normal name resolution to reject such code. Should I drop this validation patch here, and instead file a new ticket?